staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package peerproxy
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog/v2"
+
+	apidiscoveryv2 "k8s.io/api/apidiscovery/v2"
+	v1 "k8s.io/api/coordination/v1"
+	schema "k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	responsewriterutil "k8s.io/apiserver/pkg/util/responsewriter"
+)
+
+const (
+	controllerName = "peer-discovery-cache-sync"
+	maxRetries     = 5
+)
+
+func (h *peerProxyHandler) RunPeerDiscoveryCacheSync(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer h.peerLeaseQueue.ShutDown()
+	defer func() {
+		err := h.apiserverIdentityInformer.Informer().RemoveEventHandler(h.leaseRegistration)
+		if err != nil {
+			klog.Warning("error removing leaseInformer eventhandler")
+		}
+	}()
+
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, h.runWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func (h *peerProxyHandler) enqueueLease(lease *v1.Lease) {
+	h.peerLeaseQueue.Add(lease.Name)
+}
+
+func (h *peerProxyHandler) runWorker(ctx context.Context) {
+	for h.processNextElectionItem(ctx) {
+	}
+}
+
+func (h *peerProxyHandler) processNextElectionItem(ctx context.Context) bool {
+	key, shutdown := h.peerLeaseQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer h.peerLeaseQueue.Done(key)
+
+	err := h.syncPeerDiscoveryCache(ctx)
+	h.handleErr(err, key)
+	return true
+}
+
+func (h *peerProxyHandler) syncPeerDiscoveryCache(ctx context.Context) error {
+	// Rebuild the peer discovery cache from available leases.
+	leases, err := h.apiserverIdentityInformer.Lister().List(h.identityLeaseLabelSelector)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return err
+	}
+
+	newCache := map[string]*peerDiscoveryInfo{}
+	for _, l := range leases {
+		discoveryInfo, err := h.addPeerDiscoveryInfoToCache(ctx, l)
+		if err != nil {
```

Comment:
If a single discovery request fails, should we really be stopping the entire operation?

Reply:
Updated to return the error at the end of the entire operation so that we still use whatever discovery info we got from existing servers to rebuild the peerDiscoveryInfoCache, and also ensure that we retry discovery since there was an err encountered.